src/psyclone/psyir/nodes/routine.py

```diff
@@ -300,11 +331,65 @@ def _refine_copy(self, other):
         :type other: :py:class:`psyclone.psyir.node.Routine`
 
         '''
+        # Removing the parent_node knowledge from the default copy method
+        # before calling the super, else in Node we set self._parent = None
+        # and attempt to remove this symbol from a symbol table that doesn't
+        # contain it.
+        self._parent_node = None
+        self._symbol = other._symbol.copy()
         super()._refine_copy(other)
         if other.return_symbol is not None:
             self.return_symbol = self.symbol_table.lookup(
                     other.return_symbol.name)
 
+    def replace_with(self, node, keep_name_in_context=True):
+        '''Removes self and its descendents from the PSyIR tree to which it
+        is connected and replaces it with the supplied node (and its
+        descendents).
+
+        The node must be a Routine (or subclass) and has the same Symbol as
+        self.
+
+        keep_name_in_context is ignored for this replace_with implementation,
+        however we keep the argument to match with the base implementation.
+
+        :param node: the node that will replace self in the PSyIR tree.
+        :type node: :py:class:`psyclone.psyir.nodes.Routine`
+        :param bool keep_name_in_context: ignored.
+
+        :raises TypeError: if the argument node is not a Routine.
+        :raises GenerationError: if this node does not have a parent.
+        :raises GenerationError: if the argument 'node' has a parent
+        :raises GenerationError: if self and node do not have the same Symbol.
+        '''
+        if not isinstance(node, Routine):
+            raise TypeError(
+                f"The argument node in method replace_with in the Routine "
+                f"class should be a Routine but found '{type(node).__name__}'."
+            )
+        if not self.parent:
+            raise GenerationError(
+                "This node should have a parent if its replace_with method "
+                "is called.")
+        if node.parent is not None:
+            raise GenerationError(
+                f"The parent of argument node in method replace_with in the "
+                f"Routine class should be None but found "
+                f"'{type(node.parent).__name__}'.")
+        if self._symbol != node._symbol:
+            raise GenerationError(
+                "The symbol of argument node in method replace_with in the "
+                "Routine class should be the same as the Routine being "
+                "replaced.")
+        # We want to not use the childrenList update information.
+        node._symbol_in_table = self._symbol_in_table
```

Comment:
`_symbol_in_table` is persistent and keeps track of whether we were able to remove the symbol when detaching the `Routine` from its parent. When we use replace then `node._symbol_in_table = True` is probably accurate (since `self` has to have a parent to be able to be replaced), but it needs to be updated in case it is then `detach`ed somehow later.

I guess alternatively we could adjust the `_parent` setter to do `self._symbol_in_table = True` if the `ValueError` is hit when removing it from its parent and that would solve this issue without a private access, but the `_symbol_in_table` does need to be correct for correct behaviour.

Reply:
Added a comment for this, but happy to adjust this if you think it would be better.

Reply:
If I understand this correctly you use this only for the replace_with case, because in that case the remove could fail (there may be calls to it), but we know it is fine because we are re-adding a Routine with the same symbol (`self._symbol != node._symbol`) check. If this is the case, could you call it

`self._routine_being_replaced = False`

and maybe in the setter just do
```
if self._routine_being_replaced == True:
   # If this comes from a replace_with, we don't want to remove the routine when the setter
   # is given the None value nor add it again when the new parent is given. The second call
   # with a not None parent marks the end of the replace with
   if parent is not None:
       self._routine_being_replaced = False
   self._parent_node = parent  # This line can also be gone if using the "update_parent_symbol_tables"
else:
   # logic without the _routine_being_replaced
```
Would this work?

Reply:
No thats not it, its primarily for the `remove` case as we don't fail to `remove` the Routine itself, only its symbol may continue to exist in the symbol table it was removed from. So if its reattached to a parent, its used to determine whether it needs to create a copy of its symbol to place in its new parent (since the old symbol is already in a symbol table elsewhere). Failing to remove the routine on `.detach()` didn't feel like a good choice.